src/rmarkdown/preview.ts

```diff
@@ -0,0 +1,267 @@
+    }
+}
+
+class RMarkdownChildStore extends vscode.Disposable {
+    private store: Set<RMarkdownChild> = new Set<RMarkdownChild>();
+
+    constructor() {
+        super((): void => {
+            for (const child of this.store) {
+                child.dispose();
+            }
+            this.store.clear();
+        });
+    }
+
+    public add(child: RMarkdownChild): void {
+        this.store.add(child);
+    }
+
+    public delete(child: RMarkdownChild): void {
+        child.dispose();
+        this.store.delete(child);
+    }
+
+    public get(uri: vscode.Uri) {
+        for (const child of this.store) {
+            if (child.uri === uri) {
+                return child;
+            }
+        }
+        return undefined;
+    }
+
+    public has(uri: vscode.Uri) {
+        for (const child of this.store) {
+            if (child.uri === uri) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    [Symbol.iterator](): Iterator<RMarkdownChild> {
+        return this.store[Symbol.iterator]();
+    }
+}
+
+export class RMarkdownPreviewManager {
+    private rPath: string;
+    private ChildStore: RMarkdownChildStore = new RMarkdownChildStore;
+    private activePreview: RMarkdownChild;
+    private rMarkdownOutput: vscode.OutputChannel = vscode.window.createOutputChannel('R Markdown');
+
+    public async init(): Promise<void> {
+        this.rPath = await getRpath(false);
+        extensionContext.subscriptions.push(this.ChildStore);
+    }
+
+    public async previewRmd(viewer: vscode.ViewColumn, uri?: vscode.Uri): Promise<void> {
+        const fileUri = uri ?? vscode.window.activeTextEditor.document.uri;
+        const fileName = fileUri.path.substring(fileUri.path.lastIndexOf('/') + 1);
+        const previewEngine: string = config().get('rmarkdown.previewEngine');
+        const currentViewColumn: vscode.ViewColumn = vscode.window.activeTextEditor.viewColumn ?? vscode.ViewColumn.Active;
+        const cmd = (
+            `${this.rPath} --silent --slave --no-save --no-restore -e "${previewEngine}('${fileUri.path}')"`
+        );
+        const reg: RegExp = this.constructRegex(previewEngine);
+
+        if (this.ChildStore.has(fileUri)) {
+            this.ChildStore.get(fileUri).panel.reveal();
+        } else {
+            await doWithProgress(async () => {
+                await this.spawnProcess(cmd, reg, previewEngine, fileName, viewer, fileUri, currentViewColumn)
+                    .catch((cp: cp.ChildProcessWithoutNullStreams) => {
+                        void vscode.window.showErrorMessage('There was an error in knitting the document. Please check the R Markdown output stream.');
+                        cp.kill('SIGKILL');
+                    }
+                    );
+            },
+                vscode.ProgressLocation.Notification,
+                `Knitting ${fileName}...`
+            );
+        }
+    }
+
+    public refreshPanel(): void {
```

Comment:
This does not work for me anymore. Whenever I click "Refresh Preview", the WebView goes to blank.

Reply:
From the WebView developer tool, it looks like the iframe has the wrong url:
```html
<iframe src="file:///Users/ken/Workspaces/test/rtest/test.Rmd" width="100%" height="100%" frameborder="0">
</body>
</html>
</iframe>
```

Reply:
https://github.com/REditorSupport/vscode-R/pull/692/files#r663525890 should resolve this.

Reply:
Thanks for picking up on this!